File: app/src/main/java/abdullah/elamien/bloodbank/ui/SearchActivity.java
package abdullah.elamien.bloodbank.ui;

import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.NavUtils;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProviders;

import com.google.firebase.firestore.FirebaseFirestore;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.List;

import javax.inject.Inject;

import abdullah.elamien.bloodbank.R;
import abdullah.elamien.bloodbank.adapters.SearchAdapter;
import abdullah.elamien.bloodbank.databinding.ActivitySearchBinding;
import abdullah.elamien.bloodbank.di.components.DaggerSearchComponent;
import abdullah.elamien.bloodbank.di.components.SearchComponent;
import abdullah.elamien.bloodbank.di.modules.SearchAdapterModule;
import abdullah.elamien.bloodbank.eventbus.SearchEvent;
import abdullah.elamien.bloodbank.models.Donor;
import abdullah.elamien.bloodbank.viewmodels.SearchViewModel;
import abdullah.elamien.bloodbank.viewmodels.factory.ViewModelFactory;

public class SearchActivity extends AppCompatActivity {


    @Inject
    FirebaseFirestore mFirestore;
    @Inject
    SearchAdapter mSearchAdapter;
    @Inject
    ViewModelFactory mViewModelFactory;
    private ActivitySearchBinding mBinding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_search);
        mBinding.setHandlers(this);
        initDagger();
        setupToolbar();
        registerEventBus();
    }

    private void initDagger() {
        SearchComponent component = DaggerSearchComponent.builder()
                .searchAdapterModule(new SearchAdapterModule(this))
                .appComponent(((App) getApplication()).getAppComponent())
                .build();
        component.inject(this);
    }

    private void setupToolbar() {
        setSupportActionBar(mBinding.toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setDisplayShowHomeEnabled(true);
            actionBar.setTitle(getString(R.string.search_activity_label));
        }
        mBinding.toolbar.setNavigationIcon(R.drawable.ic_nav_back);
    }


    private void registerEventBus() {
        EventBus.getDefault().register(this);
    }

    @Override
    protected void onStop() {
        super.onStop();
        unregisterEventBus();
    }

    private void unregisterEventBus() {
        EventBus.getDefault().unregister(this);
    }

    @Subscribe
    public void onEvent(SearchEvent event) {
        if (event.getSearchEventMessage().toLowerCase().equals("fail")) {
            showError(getString(R.string.error_msg));
        }
    }

    private void showError(String s) {
        Toast.makeText(this, s, Toast.LENGTH_SHORT).show();
    }


    public void onSearchDonorsButtonClick(View view) {
        if (isChipsSelected()) {
            searchForDonors();
        }
    }

    private boolean isChipsSelected() {
        return isAddressChipSelected() && isBloodTypeChipSelected();
    }

    private void searchForDonors() {
        SearchViewModel viewModel = ViewModelProviders.of(this, mViewModelFactory).get(SearchViewModel.class);
        viewModel.getSearchResult(getChipAddress(), getBloodType()).observe(this, this::showSearchResults);
    }

    private void showSearchResults(List<Donor> donors) {
        mSearchAdapter.setDonors(donors);
        mBinding.searchDonorsRecyclerView.setAdapter(mSearchAdapter);
    }

    private String getBloodType() {
        switch (mBinding.bloodTypeChipGroup.getCheckedChipId()) {
            case R.id.aPlusChip:
                return getString(R.string.a_plus_blood_type_chip_label);
            case R.id.aMinusChip:
                return getString(R.string.a_minus_blood_type_chip_label);
            case R.id.bPlusChip:
                return getString(R.string.b_plus_blood_type_chip_label);
            case R.id.bMinusChip:
                return getString(R.string.b_minus_blood_type_chip_label);
            case R.id.oPlusChip:
                return getString(R.string.o_plus_blood_type_chip_label);
            case R.id.oMinusChip:
                return getString(R.string.o_minus_blood_type_chip_label);
            case R.id.abPlusChip:
                return getString(R.string.ab_plus_blood_type_chip_label);
            case R.id.abMinusChip:
                return getString(R.string.ab_minus_blood_type_chip_label);
            default:
                return "Not available";
        }
    }

    private String getChipAddress() {
        switch (mBinding.addressChipGroup.getCheckedChipId()) {
            case R.id.arishChip:
                return getString(R.string.alarish_chip_label);
            case R.id.bearAlAbdChip:
                return getString(R.string.bear_al_abd_chip_label);
            case R.id.nekhelChip:
                return getString(R.string.nekhel_chip_label);
            case R.id.sheikhZouidChip:
                return getString(R.string.sheikh_zould_chip_label);
            case R.id.rafahChip:
                return getString(R.string.rafah_chip_label);
            case R.id.hasanaChip:
                return getString(R.string.hasana_chip_label);
            default:
                return "Not available";
        }
    }

    private boolean isBloodTypeChipSelected() {
        return mBinding.bloodTypeChipGroup.getCheckedChipId() != View.NO_ID;
    }


    private boolean isAddressChipSelected() {
        return mBinding.addressChipGroup.getCheckedChipId() != View.NO_ID;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                NavUtils.navigateUpFromSameTask(this);
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/abdullah/elamien/bloodbank/ui/SearchActivity.kt
package abdullah.elamien.bloodbank.ui

import abdullah.elamien.bloodbank.R
import abdullah.elamien.bloodbank.adapters.SearchAdapter
import abdullah.elamien.bloodbank.databinding.ActivitySearchBinding
import abdullah.elamien.bloodbank.eventbus.SearchEvent
import abdullah.elamien.bloodbank.models.Donor
import abdullah.elamien.bloodbank.viewmodels.SearchViewModel
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NavUtils
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.koin.android.ext.android.inject

class SearchActivity : AppCompatActivity() {

    private val mSearchViewModel by inject<SearchViewModel>()
    private var mBinding: ActivitySearchBinding? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_search)
        mBinding!!.handlers = this
        setupToolbar()
        registerEventBus()
    }

    private fun setupToolbar() {
        setSupportActionBar(mBinding!!.toolbar)
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setDisplayShowHomeEnabled(true)
            actionBar.title = getString(R.string.search_activity_label)
        }
        mBinding!!.toolbar.setNavigationIcon(R.drawable.ic_nav_back)
    }


    private fun registerEventBus() {
        EventBus.getDefault().register(this)
    }

    override fun onStop() {
        super.onStop()
        unregisterEventBus()
    }

    private fun unregisterEventBus() {
        EventBus.getDefault().unregister(this)
    }

    @Subscribe
    fun onEvent(event: SearchEvent) {
        if (event.searchEventMessage.toLowerCase() == "fail") {
            showError(getString(R.string.error_msg))
        }
    }

    private fun showError(s: String) {
        Toast.makeText(this, s, Toast.LENGTH_SHORT).show()
    }


    fun onSearchDonorsButtonClick(view: View) {
        if (isChipsSelected) {
            searchForDonors()
        }
    }

    private fun searchForDonors() {
        mSearchViewModel.getSearchResult(chipAddress, bloodType).observe(this, Observer<List<Donor>> { this.showSearchResults(it) })
    }

    private val isChipsSelected: Boolean
        get() = isAddressChipSelected && isBloodTypeChipSelected

    private val bloodType: String
        get() {
            return when (mBinding!!.bloodTypeChipGroup.checkedChipId) {
                R.id.aPlusChip -> getString(R.string.a_plus_blood_type_chip_label)
                R.id.aMinusChip -> getString(R.string.a_minus_blood_type_chip_label)
                R.id.bPlusChip -> getString(R.string.b_plus_blood_type_chip_label)
                R.id.bMinusChip -> getString(R.string.b_minus_blood_type_chip_label)
                R.id.oPlusChip -> getString(R.string.o_plus_blood_type_chip_label)
                R.id.oMinusChip -> getString(R.string.o_minus_blood_type_chip_label)
                R.id.abPlusChip -> getString(R.string.ab_plus_blood_type_chip_label)
                R.id.abMinusChip -> getString(R.string.ab_minus_blood_type_chip_label)
                else -> "Not available"
            }
        }

    private val chipAddress: String
        get() {
            return when (mBinding!!.addressChipGroup.checkedChipId) {
                R.id.arishChip -> getString(R.string.alarish_chip_label)
                R.id.bearAlAbdChip -> getString(R.string.bear_al_abd_chip_label)
                R.id.nekhelChip -> getString(R.string.nekhel_chip_label)
                R.id.sheikhZouidChip -> getString(R.string.sheikh_zould_chip_label)
                R.id.rafahChip -> getString(R.string.rafah_chip_label)
                R.id.hasanaChip -> getString(R.string.hasana_chip_label)
                else -> "Not available"
            }
        }

    private val isBloodTypeChipSelected: Boolean
        get() = mBinding!!.bloodTypeChipGroup.checkedChipId != View.NO_ID


    private val isAddressChipSelected: Boolean
        get() = mBinding!!.addressChipGroup.checkedChipId != View.NO_ID

    private fun showSearchResults(donors: List<Donor>) {
        val adapter = SearchAdapter(this)
        adapter.setDonors(donors)
        mBinding!!.searchDonorsRecyclerView.adapter = adapter
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> NavUtils.navigateUpFromSameTask(this)
        }
        return super.onOptionsItemSelected(item)
    }
}
